page_iterator.go

```diff
@@ -0,0 +1,203 @@
+//          return true
+//      }
+//      err := pageIterator.Iterate(callbackFunc)
+func (pI *PageIterator) Iterate(callback func(pageItem interface{}) bool) error {
+	for pI.currentPage != nil {
+		keepIterating := pI.enumerate(callback)
+
+		if !keepIterating {
+			// Callback returned false, stop iterating through pages.
+			return nil
+		}
+
+		err := pI.next()
+		if err != nil {
+			return err
+		}
+		pI.pauseIndex = 0 // when moving to the next page reset pauseIndex
+	}
+
+	return nil
+}
+
+// SetHeaders provides headers for requests made to get subsequent pages
+//
+// Headers in the initial request -- request to get the first page -- are not included in subsequent page requests.
+func (pI *PageIterator) SetHeaders(headers map[string]string) {
+	pI.headers = headers
+}
+
+func (pI *PageIterator) hasNext() bool {
+	if pI.currentPage == nil || pI.currentPage.getNextLink() == nil {
+		return false
+	}
+	return true
+}
+
+func (pI *PageIterator) next() error {
+	nextPage, err := pI.getNextPage()
+	if err != nil {
+		return err
+	}
+
+	pI.currentPage = nextPage
+	return nil
+}
+
+func (pI *PageIterator) getNextPage() (*PageResult, error) {
+	if pI.currentPage.getNextLink() == nil {
+		return nil, nil
+	}
+
+	nextLink, err := url.Parse(*pI.currentPage.getNextLink())
+	if err != nil {
+		return nil, errors.New("Parsing nextLink url failed")
+	}
+
+	requestInfo := abstractions.NewRequestInformation()
+	requestInfo.Method = abstractions.GET
+	requestInfo.SetUri(*nextLink)
+	requestInfo.Headers = pI.headers
+
+	res, err := pI.reqAdapter.SendAsync(*requestInfo, pI.constructorFunc, nil)
+	if err != nil {
+		return nil, errors.New("Fetching next page failed")
+	}
+
+	return convertToPage(res)
+}
+
+func (pI *PageIterator) enumerate(callback func(item interface{}) bool) bool {
+	keepIterating := true
+
+	if pI.currentPage == nil {
+		return false
+	}
+
+	pageItems := pI.currentPage.getValue()
+	if pageItems == nil {
+		return false
+	}
+
+	if pI.pauseIndex >= len(pageItems) {
+		return false
+	}
+
+	// start/continue enumerating page items from  pauseIndex.
+	// this makes it possible to resume iteration from where we paused iteration.
+	for i := pI.pauseIndex; i < len(pageItems); i++ {
+		keepIterating = callback(pageItems[i])
+
+		if !keepIterating {
+			// Callback returned false, pause! stop enumerating page items. Set pauseIndex so that we know
+			// where to resume from.
+			// Resumes from the next item
+			pI.pauseIndex = i + 1
+			break
+		}
+	}
+
+	return keepIterating
+}
+
+func convertToPage(response interface{}) (*PageResult, error) {
+	ref := reflect.ValueOf(response).Elem()
+	value := ref.FieldByName("value")
+	value = reflect.NewAt(value.Type(), unsafe.Pointer(value.UnsafeAddr())).Elem()
+	if value.IsNil() {
+		return nil, errors.New("value property missing in response object")
+	}
+
+	nextLink := ref.FieldByName("nextLink")
+	nextLink = reflect.NewAt(nextLink.Type(), unsafe.Pointer(nextLink.UnsafeAddr())).Elem()
```

Comment:
add 2 nil checks

Reply:
It's okay of the nextLink is Nil, it means we don't have a next page to get.

Reply:
but the field itself?